wikimedia/WikiContrib
# [un-mergeable] Response time research

**This PR is the result of the research done about #58 and #110**
**This research is divided into two:**
 * Asyncio based solution
 * Bitergia dev tool console based solution

to read about the second part, skip to the bottom



**Asyncio Based Solution**
**What it addresses**
 * Finding out how long it takes to fetch the contributions of a single user on average
 * Finding out how long it takes to fetch the contributions of a list of users

**HOW TO TEST**
 * clone my copy of this repo
 * Go to backend folder and enable your virtual env
 * Do the required setup
 * Switch to response_time_research branch
 * Start server for backend and front-end and try uploading a csv with the desired number 
   of usernames
If it doesn't redirect immediately, open your network tab and copy the hash in the last URL visited, then open a new tab and visit (localhost:3000/exp/hash)

 
![list_view](https://user-images.githubusercontent.com/40905613/77235710-6c19b500-6bb8-11ea-9109-622f13a5fb56.JPG)

**My Findings**
It seems like there is a cost associated with every additional user data retrieval. By that, I mean that the relationship between the number of users and the time taken to fetch their data is exponential. This is because of phabricator API rate limits. 
From my findings, the real problem are the two Phabricator endpoints. Gerrit always returns a response in like two seconds tops even when you are fetching like 20 users data.

**Proposed solution**
It will be very beneficial if we build a caching mechanism to limit the network hit to the phab API endpoints. You can go to issue #110  to read about how I think this can be done





**Bitergia Dev Tool Console Based Solution**

@rammanoj proposed another possible way of solving this issue by scraping bitergia dev tool console and provided the starter snippet for achieving this.
I have modified this research PR to include the research work I did based on his recommendation.

**What I did**
 * improved on his snippet to read queries from json, post the queries to bitergia dev tool console, 
  retrieve the result and store it in another json file


**HOW TO TEST**
 * clone my copy of this repo
 * Go to backend folder and enable your virtual env
 * Do the required setup (most importantly by running pip install -r requirements.txt)
 * Switch to response_time_research branch
 * Download chromedriver for your system and paste the executable inside bitergia_scraping folder 
  in the backend path of the project
 * Inside the bitergia_folder, open console and run "python bitergia_scraping_script.py"

Note that you will need to install openssl on your system (if you are not running windows) for this to work, since the seleniumwire package needed for this implementation requires it.


**Findings**
Minus the time it takes to properly startup chrome driver (this varies depending on system power), the query latency seems to be very short.
If we decide to take this method though, we might have to change the implementation if bitergia ever decides to tweak their code since this is not the approved way.
I also found that the results returned from the queries are paginated to ten items by default though I couldn't see any URL to go to prev and next pagination.

**Challenges to overcome before this can become viable**
 * Find a proper documentation detailing how to make proper queries on bitergia dev tool console.
 * Find a solution to the pagination problem as that can significantly increase the query latency.

**Please Note That This is work in progress as I will continue making changes based on future research findings. Anybody is free to contribute to this as well**
@srish @rammanoj
